### PR TITLE
Propagate Replicate-specific trace options in W3C tracestate

### PR DESCRIPTION
**Pull request by nickstenning**
This sets us up to propagate our own Replicate-specific trace options in the standard trace state object used by OpenTelemetry. This means that it will automatically be propagated by all the standard OTel instrumentation libraries.

It adds two new functions to the telemetry package:

1. `TraceOptionsFromContext`, which extracts the trace options from the current span context if present.

2. `WithTraceOptions`, which adds trace options to the current context.

There is one additional convenience function, `WithFullTrace`, which enables "full trace" mode. This sets the detail level trace option to "full" and the sample mode trace option to "always".

The second commit adds a `SpanProcessor`, `TraceOptionsProcessor`, which automatically adds the "meta.replicate.disable_sampling" span attribute whenever the SampleMode trace option is set to "always".

Together with the first change, this means that any Go application in our stack will automatically add this attribute to its spans when the tracestate contains our trace options keys. One caveat is that this only applies for spans started after the tracestate has been added to the span context.

### CODE REVIEW

File: telemetry/tracestate.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+const (
+	SampleModeNever SampleMode = iota - 1
+	SampleModeDefault
+	SampleModeAlways
+)
+
+// Don't forget to rerun `go generate ./...` if you add or remove values from
+// the DetailLevel or SampleMode enums.
+//
+//go:generate stringer -type=DetailLevel,SampleMode -output=tracestate_string.go
```
Comment (philandstuff) [inline telemetry/tracestate.go:29]: is it worth adding a check to script/lint or script/format (and CI) that `go generate` is up to date?

Comment (nickstenning) [reply telemetry/tracestate.go:29]: Good idea!

File: telemetry/processors.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+// Check TraceOptionsProcessor implements SpanProcessor
+var _ trace.SpanProcessor = new(TraceOptionsProcessor)
+
+// TraceOptionsProcessor handles any custom span handling related to our own
+// TraceOptions. At the moment it takes care of adding attributes to the span
+// that control sampling and detail level.
+type TraceOptionsProcessor struct {
+	Next trace.SpanProcessor
```
Comment (philandstuff) [inline telemetry/processors.go:17]: Is Next deliberately public?

Comment (nickstenning) [reply telemetry/processors.go:17]: So we don't need a constructor?

**Review by philandstuff — APPROVED** [review]
looks fine, nothing blocking

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -37,7 +37,9 @@ jobs:
       with:
         go-version-file: go.mod
         cache: false # for this job we rely on golangci-lint-action's cache
-    - name: Lint
+    - name: Lint (other)
+      run: script/lint
+    - name: Lint (golangci-lint)
       uses: golangci/golangci-lint-action@v3
```
[the remark this answers is not in the document]
Comment (nickstenning) [reply .github/workflows/ci.yaml:43]: The action handles adding lint stuff as comments on the PR.

**Pull request merged**